Feat: Genre Implementation

Jamie001129 opened the pull request:
I redid the genre implementation. This time I tried to use all the stuff from the author's repo for accurate reproduction, with just a few modifications to fit our repo. The GenRe models I trained on my Mac and in my environment (Python 3.12, PyTorch "a kind-of-recent version") achieved better reproduction results but are, unfortunately, incompatible with our Python 3.7 environment, and therefore useless. T_T

Now the models are hosted on huggingface (https://huggingface.co/jamie250/genrereproduce) and the reproduce.py can fetch them. We need to add the huggingface-hub library to our environment to support fetching models from huggingface so I modified the requirements-dev.txt as well.

Paper:
FROM SEARCH TO SAMPLING: GENERATIVE MODELS FOR ROBUST ALGORITHMIC RECOURSE
By Prateek Garg*, Lokesh Nagalapatti, Sunita Sarawagi
ICLR 2025

Results (Table 2, Adult dataset, shown as "mine/original"):
Cost: 0.70/0.69
Val: 0.93/1.00
LOF: 0.90/0.98
Score: 1.78/1.93

Potential Factors Affecting Metrics:
1 Environment, especially PyTorch version differences
2 Batch size for transformer training 8192 -> 1024

Commit:
the third time is the "charm", so what about the the fourth?

zkhotanlou commented on methods/catalog/genre/model.py:
```diff
@@ -0,0 +1,228 @@
+    
+    def get_counterfactuals(self, factuals: pd.DataFrame) -> pd.DataFrame:
+        """
+        Generate counterfactuals
```
The purpose of this method is to calculate counterfactuals for the input factuals, not to act as a fake interface. The returned counterfactuals should be specific to the given factual input, not always a fixed output for the entire X_test to artificially reproduce the results.

zkhotanlou commented on methods/catalog/genre/reproduce.py:
```diff
@@ -0,0 +1,388 @@
+    
+    # Initialize GenRe recourse module
+    rec_module = GenReOriginal(
+        pair_model=pair_model,
```
The recourse method should be able to work with ModelCatalog, which itself works with DataCatalog, this implementation doesn't align with the repo structure.

zkhotanlou commented on methods/catalog/genre/reproduce.py:
```diff
@@ -0,0 +1,388 @@
+
+
+if __name__ == "__main__":
+    main()
```
I tried to run this script, but it doesn't run successfully and gets a ModuleNotFoundError.

Commit:
solve pytest error

Jamie001129 committed:
update model and reproduce to follow the repo's structure

Commit:
fix import errors

Commit:
continue sloving the import issues, add library. prefix wherever necessay

Commit:
fix expected range

Commit:
delete previous saved results

Jamie001129 committed:
Merge branch 'charmlab:main' into feature/add-genre-v4

Commit:
Fix pre-commit issues

zkhotanlou committed:
Fix: moduleNotFound error

zkhotanlou committed:
fix: lint test.py

zkhotanlou committed:
Fix: moduleNotFound error

zkhotanlou committed:
Fix: moduleNotFound error

zkhotanlou commented on methods/catalog/genre/reproduce.py:
```diff
@@ -0,0 +1,318 @@
+    # 5. Initialize GenRe using our wrapper
+    print("Initializing GenRe...")
+    genre = GenRe(
+        mlmodel=ann_clf,  # Pass ANN directly
```
The recourse method still does not align with the repository’s ModelCatalog and DataCatalog structure. Even if the current implementation successfully reproduces the original paper’s results, it cannot yet operate within the benchmark pipeline. After reproduction, the method must be runnable on all existing datasets and models in the benchmark suite. In its current form, the benchmark pipeline would fail.

Jamie001129 committed:
Support both author models and repo ModelCatalog in GenRe

Commit:
start to add agenre that fits the repo

Commit:
let model.py load the trasnformer since the transformer is part of the Genre method
Add a fucntion in reproduce.py to test compatibility(working on it)
TODO add data add model if current ones don't work

Commit:
How to get the catmask?

Commit:
update model.py so it cant calculate mask based on data

Jamie001129 committed:
add genre_adult dataset to datacatalog

Jamie001129 committed:
fix reproduyce2.py to add test_compatibility calls with correct parameters

Commit:
fix some bug but there's more to fix

Jamie001129 committed:
add checkdimensions script

Commit:
create wrapper in model.py to change the categorical output of repo's mlp to binary output
clean genre folder

Commit:
fix precommit hook

zkhotanlou committed:
Merge branch 'main' into feature/add-genre-v4

zkhotanlou reviewed (commented):
After addressing the comments, please add the results of you implemented method on benchmark datasets and models to run_experiments.py and result.csv

Jamie001129 committed:
Merge branch 'charmlab:main' into feature/add-genre-v4

Jamie001129 commented on data/catalog/_data_main/process_data/process_genre_adult_data.py:
But the data already exists it in the genre directory. These changes are for accessing my data through Datacatalog

Jamie001129 committed:
Add pytest parametrize decorator to test_compatibility for automated testing

Jamie001129 committed:
Merge branch 'main' into feature/add-genre-v4

Commit:
the genre adult in the yaml is gone?????
So i add it back

Commit:
remove comments

Commit:
genre_adult in mlmodel_catalog.yaml is also gone???????
add it back!

Commit:
the is something wrong with the uci_credit part
fixed it

zkhotanlou committed:
Fix: precommit hooks

zkhotanlou committed:
Fix: conflicts

zkhotanlou commented:
This is an implementation of the "Genre"[1] recourse method. The level of reproduction is on level1 as the unit tests check the implementation could reproduce results reported in the paper for the Adult dataset on MLP model, and also check to be compatible with the repo's structure.

[1] Garg, P., Nagalapatti, L., & Sarawagi, S. (2025). From Search To Sampling: Generative Models For Robust Algorithmic Recourse. arXiv preprint arXiv:2505.07351.
